### PR TITLE
[Orchestrator]  add new check to collect ecs tasks

### PR DESCRIPTION
**Pull request by kangyili**
### What does this PR do?

**This PR requires**

1. https://github.com/DataDog/datadog-agent/pull/21836
2. https://github.com/DataDog/datadog-agent/pull/23253
3. https://github.com/DataDog/datadog-agent/pull/23823
4. https://github.com/DataDog/agent-payload/pull/281

#### New check `orchestrator_ecs` 

* Add a new check `orchestrator_ecs` in core checks to collect running tasks from ecs-ec2 and ecs-fargate.
* The new check is controlled by `DD_ORCHESTRATOR_EXPLORER_ENABLED` and `DD_ORCHESTRATOR_EXPLORER_ECS_COLLECTION_ENABLED`
* It uses `Workloadmeta` for pulling tasks in every run https://github.com/DataDog/datadog-agent/blob/f67c3d667e00dcf6350a9d80a7075998e57ad9e3/pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go#L57
* The payloads are sent to orchestrator endpoints

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by cswatt — APPROVED** [review]
release notes approved by docs

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by agent shared components

File: go.mod [hunk 1]
```diff
@@ -592,7 +592,7 @@ require github.com/lorenzosaino/go-sysctl v0.3.1
 
 require (
 	github.com/DATA-DOG/go-sqlmock v1.5.0
-	github.com/DataDog/agent-payload/v5 v5.0.100
+	github.com/DataDog/agent-payload/v5 v5.0.103-0.20240102141342-488df63c5468
```
Comment (kangyili) [inline go.mod:595]: It needs to be updated when https://github.com/DataDog/agent-payload/pull/281 is merged

File: pkg/orchestrator/model/types.go [hunk 2]
```diff
@@ -77,6 +77,8 @@ const (
 	K8sVerticalPodAutoscaler = 22
 	// K8sHorizontalPodAutoscaler represents a Kubernetes Horizontal Pod Autoscaler
 	K8sHorizontalPodAutoscaler = 23
+	// ECSTask represents an ECS Task
+	ECSTask = 24
```
Comment (xlucas) [inline pkg/orchestrator/model/types.go:81]: should we change this to not disrupt the k8s resource increment?
maybe start in another range (eg 100?)

**Review by xlucas — CHANGES_REQUESTED** [review]

File: pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go [hunk 3]
```diff
@@ -229,11 +229,11 @@ func (cb *CollectorBundle) importCollectorsFromDiscovery() bool {
 
 	collectors, err := discovery.NewAPIServerDiscoveryProvider().Discover(cb.inventory)
 	if err != nil {
-		_ = cb.check.Warnf("Collector discovery failed: %s", err)
+		_ = cb.check.Warnf("K8sCollector discovery failed: %s", err)
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/collector_bundle.go:232]: K8sCollector -> Kubernetes collector

File: pkg/collector/corechecks/cluster/orchestrator/collectors/collector.go [hunk 4]
```diff
@@ -105,6 +65,9 @@ type CollectorRunConfig struct {
 	Config                      *config.OrchestratorConfig
 	MsgGroupRef                 *atomic.Int32
 	OrchestratorInformerFactory *OrchestratorInformerFactory
+
+	// used by ecs collector
+	WorkloadmetaStore workloadmeta.Component
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/collectors/collector.go:70]: I think it'd be nice to create and embed a K8sCollectorRunConfig and ECSCollectorRunConfig struct to have separation of common config and orchestrator-specific configs.

File: pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		Result:             processResult,
+		ResourcesListed:    len(list),
+		ResourcesProcessed: processed,
+	}
+
+	return result, nil
+}
+
+// FetchContainers fetches the containers from workloadmeta store for a given task.
+func (t *TaskCollector) FetchContainers(rcfg *collectors.CollectorRunConfig, task *workloadmeta.ECSTask) transformers.TaskWithContainers {
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go:99]: nit: function shouldn't be exposed `FetchContainers` -> `fetchContainers`

File: pkg/orchestrator/model/types.go (hunk 2 above)
Comment (kangyili) [reply pkg/orchestrator/model/types.go]: now it's 150 https://github.com/DataDog/datadog-agent/blob/7bac94cf2d33a37f20e5c7499c742b1d499fa584/pkg/orchestrator/model/types.go#L81
,matching with https://github.com/DataDog/agent-payload/pull/281/files#diff-69b34851fbf6263111121260c46c57084a1fe6e34e9b750a5a673694f2dfe1b0R768

File: pkg/collector/corechecks/cluster/orchestrator/processors/common/base.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	model "github.com/DataDog/agent-payload/v5/process"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/cluster/orchestrator/processors"
+)
+
+// BaseHandlers implements basic Handlers
+type BaseHandlers struct{}
+
+// BeforeCacheCheck is a handler called before cache check.
+func (BaseHandlers) BeforeCacheCheck(_ processors.ProcessorContext, _, _ interface{}) (skip bool) {
```
[the remark this answers is not in the document]
Comment (kangyili) [reply pkg/collector/corechecks/cluster/orchestrator/processors/common/base.go]: It makes sense, I've reverted those change
https://github.com/DataDog/datadog-agent/pull/22060/files#diff-62e0377b239aa3f3652dc400fb04f299043763f7fefaadd5f879c35076ca7843

**Comment by kangyili** [comment]
I created a new PR only updates code structure

File: go.mod [hunk 7]
```diff
@@ -604,7 +604,7 @@ require github.com/lorenzosaino/go-sysctl v0.3.1
 
 require (
 	github.com/DATA-DOG/go-sqlmock v1.5.0
-	github.com/DataDog/agent-payload/v5 v5.0.106
+	github.com/DataDog/agent-payload/v5 v5.0.108-0.20240318102635-7d4963c15425
```
Comment (fisherevans) [inline go.mod:607]: just a reminder

File: pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		ResourcesProcessed: processed,
+	}
+
+	return result, nil
+}
+
+// fetchContainers fetches the containers from workloadmeta store for a given task.
+func (t *TaskCollector) fetchContainers(rcfg *collectors.CollectorRunConfig, task *workloadmeta.ECSTask) transformers.TaskWithContainers {
+	ecsTask := transformers.TaskWithContainers{
+		Task:       task,
```
Comment (fisherevans) [inline pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go:101]: will task here be reset to the last value of task in the parent loop, for all entries? The ptr is reused in the for.. I'm not sure if that cascades into this function call.

**Review by fisherevans — APPROVED** [review]

File: pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go (hunk 8 above)
Comment (kangyili) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go:102]: It won't but I could never figure out this looping variable issue, so I updated it to make sure it's always correct, and updated the integration test to include this part
https://github.com/DataDog/datadog-agent/pull/22060/commits/9fd8b7a3d7ac7cdf81ad97a36cf575b91067344e

**Review by vickenty — APPROVED** [review]

File: pkg/collector/corechecks/cluster/orchestrator/transformers/ecs/task.go [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+		Containers:              extractECSContainer(task.Containers),
+	}
+
+	tags, err := tagger.Tag(fmt.Sprintf("ecs_task://%s", task.Task.EntityID.ID), collectors.HighCardinality)
+	if err != nil {
+		log.Debugf("Could not retrieve tags for task: %s", err.Error())
+	}
+
+	taskModel.Tags = tags
+	taskModel.EcsTags = toTags(task.Task.Tags)
```
Comment (xlucas) [inline pkg/collector/corechecks/cluster/orchestrator/transformers/ecs/task.go:65]: Is there any normalization needed for those tags?

File: pkg/config/env/environment_container_features.go [hunk 10]
```diff
@@ -23,6 +23,8 @@ const (
 	EKSFargate Feature = "eksfargate"
 	// KubeOrchestratorExplorer can be enabled
 	KubeOrchestratorExplorer Feature = "orchestratorexplorer"
+	// ECSOrchestratorExplorer can be enabled
+	ECSOrchestratorExplorer Feature = "ecsorchestratorexplorer"
```
Comment (xlucas) [inline pkg/config/env/environment_container_features.go:27]: Is it intended to not have non alphanumeric characters here or can we have something clearer like `orchestratorexplorer_ecs` (and maybe in near future rename to `orchestratorexplorer_k8s`)?

File: pkg/collector/corechecks/orchestrator/ecs/ecs.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+type Check struct {
+	core.CheckBase
+	WorkloadmetaStore workloadmeta.Component
+	sender            sender.Sender
+	config            *oconfig.OrchestratorConfig
+	collectors        []collectors.Collector
+	groupID           *atomic.Int32
+
+	// isECSCollectionEnabledFunc is used for testing
+	isECSCollectionEnabledFunc func() bool
```
Comment (xlucas) [inline pkg/collector/corechecks/orchestrator/ecs/ecs.go:42]: could we instead mutate the check config in the test case?

**Review by xlucas — APPROVED** [review]

File: pkg/config/env/environment_container_features.go (hunk 10 above)
Comment (kangyili) [reply pkg/config/env/environment_container_features.go]: https://github.com/DataDog/datadog-agent/blob/7225ef1ca869e403d455ac5c66e3f3086a0c3fb3/pkg/config/env/environment_container_features.go#L25-L28

File: pkg/collector/corechecks/orchestrator/ecs/ecs.go (hunk 11 above)
Comment (kangyili) [reply pkg/collector/corechecks/orchestrator/ecs/ecs.go:50]: I removed this one

File: pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+	ctx := &processors.ECSProcessorContext{
+		BaseProcessorContext: processors.BaseProcessorContext{
+			Cfg:              rcfg.Config,
+			MsgGroupID:       rcfg.MsgGroupRef.Inc(),
+			NodeType:         t.metadata.NodeType,
+			ManifestProducer: t.metadata.IsManifestProducer,
+		},
+	}
+	if len(list) > 0 {
```
[the remark this answers is not in the document]
Comment (kangyili) [reply pkg/collector/corechecks/cluster/orchestrator/collectors/ecs/task.go]: I move this part to the beginning of the check 
https://github.com/DataDog/datadog-agent/blob/7225ef1ca869e403d455ac5c66e3f3086a0c3fb3/pkg/collector/corechecks/orchestrator/ecs/ecs.go#L51-L54

if it fails, the check will be sipped

File: pkg/collector/corechecks/cluster/orchestrator/transformers/ecs/task.go (hunk 9 above)
Comment (kangyili) [reply pkg/collector/corechecks/cluster/orchestrator/transformers/ecs/task.go:65]: I'm not sure if we should do the normalisation here as the tag standard of ecs and datadog are different, changing tags can confuse user. We can probably leave it as it is for now

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**